Feature idea: modifier/node interpolator

ansh7432 opened the pull request:
##  Add nodes string interpolator for mixing text and dynamic content
## Summary
This PR implements the nodes string interpolator feature requested in #401. It enables mixing static text with dynamic content (such as signals and modifiers) in a clean, intuitive way.

## Implementation Details
The implementation adds a StringContext extension method called nodes that allows for string interpolation with various types:

Signal[F, String] and Signal[F, ?] for reactive values
Modifier[F, El, ?] for existing modifiers
Product types (tuples and case classes)
Effect types like F[String]
Standard values (using toString)
Each interpolated string becomes a series of text nodes and dynamic content, preserving the reactivity of signals while providing a more ergonomic API.

## Example Usage
 ##Benefits
 Improved readability: Code is more concise and follows natural expression patterns
Maintained reactivity: Only the signal parts update when their values change
 Type flexibility: Works with various types beyond just Signals
 Intuitive syntax: Familiar string interpolation pattern from standard Scala

ansh7432 committed:
node/interpolator

ansh7432 committed:
node/interpolator

ansh7432 committed:
node/interpolator

ansh7432 committed:
updated

ansh7432 committed:
fixed

ansh7432 committed:
fixed

ansh7432 committed:
fixed nodeinterpolator

ansh7432 committed:
fixed sbt

ansh7432 committed:
fix-sbt

ansh7432 commented:
Hii @kubukoz @armanbilge please review the PR i had made the changes

armanbilge commented:
Thanks for the PR! To test it works, can you update the TodoMvc demo app to use this new feature?

ansh7432 commented:
> Thanks for the PR! To test it works, can you update the TodoMvc demo app to use this new feature?

Yeah sure making these changes.. :)

ansh7432 committed:
test added

ansh7432 committed:
added test

ansh7432 committed:
without tests

ansh7432 committed:
add test

ansh7432 committed:
added

ansh7432 committed:
fix scala

ansh7432 commented:
Hii @armanbilge @kubukoz getting some error while implementing the test for nodeinterpolator please go through the code and review it

kubukoz commented:
Can you push the code with the broken test? The last CI run is green

ansh7432 committed:
try to add nodesinterpolator at label

ansh7432 commented:
Hi @kubukoz try to add test in todomvc file adding nodesinterpolator please check and review when you have time                         I've made these changes:

Added the import: import calico.html.nodes
Updated the label in TodoItem to use the interpolator:
label(nodes"${todo.map(_.map(_.text).getOrElse(""))}") 

When compiling, I'm seeing errors related to the nodes string interpolator. Initially, I tried using it with type parameters like nodes[IO]"...", but that caused syntax errors. When removing the type parameters, it still fails to compile.

armanbilge commented:
@ansh7432 I think your implementation is quite complex. I recommend to start from the initial proposal in https://github.com/armanbilge/calico/issues/401#issuecomment-2381601846 and support additional features one-by-one, instead of all at the same time. That will make it easier to debug.

ansh7432 committed:
added nodesinterpolator in todoMVC

ansh7432 committed:
fixed todoMvc nodesinterpolator

ansh7432 committed:
fixed todomvc imports

ansh7432 commented:
Hii @armanbilge @kubukoz implemented nodesinterpolator test in todoMvc and its working fine please check if needs more changes :)

armanbilge commented:
> and its working fine

Hmm, are you sure? I just gave it a try, and the TodoMVC app seems to be broken, I am unable to add any todos 😕 

For reference, here is the working example: https://armanbilge.github.io/calico/todomvc/

ansh7432 committed:
added nodes checked

ansh7432 commented:
![image](https://github.com/user-attachments/assets/34730533-7481-4b67-8240-83456728ee3c)
hii @armanbilge I apologize, please check it once more its working in my local machine now..

ansh7432 commented:
Hi @armanbilge @kubukoz gentle reminder to review this 😊

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -0,0 +1,146 @@
+              Resource.pure[IO, Unit](())
+
+            case DynamicContent(signal) =>
+              // Use the counter instead of UUID
+              val spanId = NodeCounter.nextId()
+              val span = org.scalajs.dom.document.createElement("span")
+              span.setAttribute("id", spanId)
```
What is the purpose of this?

ansh7432 replied:
I had removed this at that time getting some compile error that's why used this now its resolved

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -0,0 +1,146 @@
+        res.flatMap { _ =>
+          content match {
+            case StaticContent(text) =>
+              val textNode = org.scalajs.dom.document.createTextNode(text)
+              element.asInstanceOf[org.scalajs.dom.Element].appendChild(textNode)
+              Resource.pure[IO, Unit](())
```
This is not pure, it should be suspended as a side-effect.

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -0,0 +1,146 @@
+  }
+
+// Add this helper method
+private def isSignal(obj: Any): Boolean =
+  obj != null && obj.getClass.getName.contains("fs2.concurrent.Signal")
```
What about other `Signal` implementations outside of the `fs2.concurrent` package?

kubukoz replied:
In general, I think the runtime class checks and casts are going to cause problems down the line - the interpolator should know the types of its inputs at compile-time.

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -0,0 +1,146 @@
+extension (sc: StringContext) {
+  // String interpolator with variadic arguments (standard string interpolator style)
+  @nowarn("msg=pattern selector should be an instance of Matchable")
+  def nodes(args: Any*): List[NodeContent] = {
```
Instead of exposing a `List[NodeContent]`, maybe we hide this implementation detail with a wrapper class `NodeInterpolator`

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -0,0 +1,146 @@
+              val spanId = NodeCounter.nextId()
+              val span = org.scalajs.dom.document.createElement("span")
+              span.setAttribute("id", spanId)
+              element.asInstanceOf[org.scalajs.dom.Element].appendChild(span)
```
Same here, the side-effects should be properly suspended.

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -0,0 +1,146 @@
+                    }
+                    .compile
+                    .drain
+                    .start
+                )(fiber => fiber.cancel)
```
This is `Concurrent#background` :)

ansh7432 replied:
didn't gett this could you please tell more :( i had raised the PR fixing all the requested reviews except this please review @armanbilge

armanbilge replied:
You can simplify this code by using the existing `background` method that does the same thing.
https://github.com/typelevel/cats-effect/blob/7102a235fd8beeeb8b13131552706c4127b0fb7a/kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala#L216-L217

ansh7432 commented:
Thanks for the review fixing all the things and again raising pr soon 😊

ansh7432 committed:
resolved

ansh7432 committed:
resolved scalafmt

ansh7432 committed:
fixed

ansh7432 committed:
fixed the background

armanbilge commented on calico/src/main/scala/calico/html/NodesInterpolator.scala:
```diff
@@ -122,20 +122,17 @@ given nodesInterpolatorModifier[E <: HtmlElement[IO]]: Modifier[IO, E, NodesInte
+                    .compile
+                    .drain
+
+                  stream.background.as(())
```
`as(())` is the definition of `void` :)

```suggestion
                  stream.background.void
```

ansh7432 replied:
apologize for this much changes will improve my PR's :( understood so many things thankyou, willing to contribute more :)

ansh7432 replied:
if needs any more changes please tell :)

ansh7432 committed:
Update calico/src/main/scala/calico/html/NodesInterpolator.scala

Co-authored-by: Arman Bilge <armanbilge@gmail.com>

armanbilge commented:
Nice work, this is much improved! I think like Jakub suggested in https://github.com/armanbilge/calico/pull/441#discussion_r1995602459, the next step is move more of the interpolation into compile-time.

> In general, I think the runtime class checks and casts are going to cause problems down the line - the interpolator should know the types of its inputs at compile-time.

ansh7432 commented:
> Nice work, this is much improved! I think like Jakub suggested in [#441 (comment)](https://github.com/armanbilge/calico/pull/441#discussion_r1995602459), the next step is move more of the interpolation into compile-time.
> 
> > In general, I think the runtime class checks and casts are going to cause problems down the line - the interpolator should know the types of its inputs at compile-time.

Yeah so looking for this now after working on this would this be able to merge ?

armanbilge commented:
Now that we've prototyped the idea, let's keep improving it in this PR until it's production-ready 💪

ansh7432 commented:
Yeah sure will working on this @armanbilge  😃

ansh7432 committed:
added macros

ansh7432 committed:
fixed

ansh7432 committed:
Apply Scalafix formatting fixes

ansh7432 committed:
Apply Scalafix formatting fixes

ansh7432 commented:
hii @armanbilge I had updated the PR please look into this.. 😊 please suggest more improvements on this 😀

ansh7432 committed:
more improvements

ansh7432 commented:
Hii @kubukoz @armanbilge please review this :)

ansh7432 committed:
more improvements
